File: repos/rappterbook-agent/python/openrappter/agents/tracer.py
"""
AgentTracer - Observability and distributed tracing for the agent framework.

Provides span-based tracing across agent executions, chains, and sub-agent calls.
Each agent invocation creates a TraceSpan. Spans link to parents via TraceContext
propagation, forming a complete trace tree across multi-agent pipelines.

Mirrors TypeScript agents/tracer.ts
"""

import uuid
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Optional


@dataclass
class TraceSpan:
    """A single span in a trace."""
    id: str
    parent_id: Optional[str]
    trace_id: str
    agent_name: str
    operation: str
    start_time: str
    end_time: Optional[str] = None
    duration_ms: Optional[int] = None
    status: str = 'running'  # 'running', 'success', 'error'
    inputs: Optional[dict] = None
    outputs: Optional[dict] = None
    data_slush: Optional[dict] = None
    error: Optional[str] = None
    tags: Optional[dict] = None


@dataclass
class TraceContext:
    """Propagated context for linking parent-child spans."""
    trace_id: str
    span_id: str
    baggage: Optional[dict] = None


class AgentTracer:
    """Span-based tracing for agent execution."""

    def __init__(self, options=None):
        options = options or {}
        self._max_spans = options.get('max_spans', 1000)
        self._record_io = options.get('record_io', True)
        self._on_span_complete = options.get('on_span_complete')
        self._active_spans = {}  # span_id -> TraceSpan
        self._completed_spans = []  # list of TraceSpan

    def start_span(self, agent_name, operation, parent_context=None, inputs=None):
        """Create and start a new span.

        Returns (span, context) tuple.
        """
        span_id = self._generate_id()
        trace_id = parent_context.trace_id if parent_context else self._generate_id()
        parent_id = parent_context.span_id if parent_context else None

        span = TraceSpan(
            id=span_id,
            parent_id=parent_id,
            trace_id=trace_id,
            agent_name=agent_name,
            operation=operation,
            start_time=datetime.now(timezone.utc).isoformat(),
            status='running',
        )

        if self._record_io and inputs is not None:
            span.inputs = self._sanitize_io(inputs)

        if parent_context and parent_context.baggage:
            span.tags = dict(parent_context.baggage)

        self._active_spans[span_id] = span

        context = TraceContext(
            trace_id=trace_id,
            span_id=span_id,
            baggage=parent_context.baggage if parent_context else None,
        )

        return span, context

    def end_span(self, span_id, result=None):
        """Complete a span. Returns the completed span or None if not found."""
        span = self._active_spans.get(span_id)
        if not span:
            return None

        end_time = datetime.now(timezone.utc)
        start_ms = datetime.fromisoformat(span.start_time).timestamp() * 1000

        span.end_time = end_time.isoformat()
        span.duration_ms = int(end_time.timestamp() * 1000 - start_ms)
        span.status = (result or {}).get('status', 'success')

        if result:
            if 'error' in result:
                span.error = result['error']
            if self._record_io and 'outputs' in result:
                span.outputs = self._sanitize_io(result['outputs'])
            if 'data_slush' in result:
                span.data_slush = result['data_slush']

        del self._active_spans[span_id]
        self._add_completed(span)

        if self._on_span_complete:
            self._on_span_complete(span)

        return span

    def get_span(self, span_id):
        """Retrieve a span by ID, searching both active and completed spans."""
        if span_id in self._active_spans:
            return self._active_spans[span_id]
        for span in self._completed_spans:
            if span.id == span_id:
                return span
        return None

    def get_trace(self, trace_id):
        """Get all spans belonging to a trace, in chronological order."""
        active = [s for s in self._active_spans.values() if s.trace_id == trace_id]
        completed = [s for s in self._completed_spans if s.trace_id == trace_id]
        all_spans = active + completed
        all_spans.sort(key=lambda s: s.start_time)
        return all_spans

    def get_active_spans(self):
        """Get all spans currently in the 'running' state."""
        return list(self._active_spans.values())

    def get_completed_spans(self, limit=None):
        """Get recently completed spans, newest first."""
        reversed_spans = list(reversed(self._completed_spans))
        if limit is not None:
            return reversed_spans[:limit]
        return reversed_spans

    def clear(self):
        """Clear all spans (active and completed)."""
        self._active_spans.clear()
        self._completed_spans = []

    def to_json(self):
        """Return a serializable summary of current tracer state."""
        completed_by_trace = {}
        for span in self._completed_spans:
            completed_by_trace.setdefault(span.trace_id, []).append(span)

        traces = []
        for trace_id, spans in completed_by_trace.items():
            sorted_spans = sorted(spans, key=lambda s: s.start_time)
            root = next((s for s in sorted_spans if s.parent_id is None), sorted_spans[0] if sorted_spans else None)
            errors = [s for s in spans if s.status == 'error']
            total_ms = sum(s.duration_ms or 0 for s in spans)

            traces.append({
                'traceId': trace_id,
                'spanCount': len(spans),
                'status': 'error' if errors else 'success',
                'rootAgent': root.agent_name if root else None,
                'startTime': root.start_time if root else None,
                'totalDurationMs': total_ms,
                'errorCount': len(errors),
            })

        traces.sort(key=lambda t: t.get('startTime') or '', reverse=True)

        return {
            'activeSpans': [self._span_to_dict(s) for s in self._active_spans.values()],
            'completedSpans': [self._span_to_dict(s) for s in reversed(self._completed_spans)],
            'traceCount': len(completed_by_trace),
            'maxSpans': self._max_spans,
            'recordIO': self._record_io,
            'traces': traces,
        }

    def _generate_id(self):
        """Generate a 16-character ID."""
        return uuid.uuid4().hex[:16]

    def _sanitize_io(self, io_dict):
        """Sanitize an IO record for safe storage."""
        if not isinstance(io_dict, dict):
            return io_dict

        sanitized = {}
        for key, value in io_dict.items():
            if key in ('_context', '_slosh_filter', '_slosh_preferences'):
                continue
            if isinstance(value, str) and len(value) > 500:
                sanitized[key] = value[:500] + '...'
            elif value is not None and isinstance(value, dict):
                sanitized[key] = '[object]'
            else:
                sanitized[key] = value
        return sanitized

    def _add_completed(self, span):
        """Append a completed span and enforce the maxSpans retention limit."""
        self._completed_spans.append(span)
        if len(self._completed_spans) > self._max_spans:
            self._completed_spans = self._completed_spans[-self._max_spans:]

    def _span_to_dict(self, span):
        """Convert a TraceSpan to a plain dict for serialization."""
        d = {
            'id': span.id,
            'parentId': span.parent_id,
            'traceId': span.trace_id,
            'agentName': span.agent_name,
            'operation': span.operation,
            'startTime': span.start_time,
            'endTime': span.end_time,
            'durationMs': span.duration_ms,
            'status': span.status,
        }
        if span.inputs is not None:
            d['inputs'] = span.inputs
        if span.outputs is not None:
            d['outputs'] = span.outputs
        if span.data_slush is not None:
            d['dataSlush'] = span.data_slush
        if span.error is not None:
            d['error'] = span.error
        if span.tags is not None:
            d['tags'] = span.tags
        return d


def create_tracer(options=None):
    """Create a new AgentTracer with the given options."""
    return AgentTracer(options)


# Global tracer singleton
global_tracer = create_tracer()

File: repos/openrappter/typescript/src/agents/tracer.ts
/**
 * AgentTracer — Observability and distributed tracing for the agent framework.
 *
 * Provides OpenTelemetry-inspired span-based tracing across agent executions,
 * chains, and sub-agent calls. Each agent invocation (execute, perform, slosh)
 * creates a TraceSpan. Spans link to parents via TraceContext propagation,
 * forming a complete trace tree across multi-agent pipelines.
 *
 * Usage:
 *   const { span, context } = tracer.startSpan('ShellAgent', 'execute', undefined, { action: 'bash' });
 *   // ... run agent ...
 *   tracer.endSpan(span.id, { status: 'success', outputs: { exitCode: 0 } });
 *
 *   const trace = tracer.getTrace(context.traceId);
 *   // trace is an array of all spans in chronological order
 */

import { randomUUID } from 'crypto';

// ── Interfaces ────────────────────────────────────────────────────────────────

export interface TraceSpan {
  /** Unique span ID */
  id: string;
  /** Parent span ID (null for root) */
  parentId: string | null;
  /** Trace ID (shared across the whole trace tree) */
  traceId: string;
  /** Agent name */
  agentName: string;
  /** Span operation name (e.g., 'execute', 'perform', 'slosh') */
  operation: string;
  /** ISO 8601 start time */
  startTime: string;
  /** ISO 8601 end time (null if still running) */
  endTime: string | null;
  /** Duration in ms (null if still running) */
  durationMs: number | null;
  /** Status */
  status: 'running' | 'success' | 'error';
  /** Input kwargs (sanitized) */
  inputs?: Record<string, unknown>;
  /** Output result summary */
  outputs?: Record<string, unknown>;
  /** Data slush flow */
  dataSlush?: {
    upstream: Record<string, unknown> | null;
    downstream: Record<string, unknown> | null;
  };
  /** Error message if status is 'error' */
  error?: string;
  /** Arbitrary tags */
  tags?: Record<string, string>;
}

export interface TraceContext {
  /** Trace ID */
  traceId: string;
  /** Current span ID (becomes parentId for child spans) */
  spanId: string;
  /** Baggage items (propagated key-value pairs) */
  baggage?: Record<string, string>;
}

export interface AgentTracerOptions {
  /** Max completed spans to retain (default: 1000) */
  maxSpans?: number;
  /** Whether to record inputs/outputs (default: true) */
  recordIO?: boolean;
  /** Custom span handler called when a span completes */
  onSpanComplete?: (span: TraceSpan) => void;
}

// ── AgentTracer ───────────────────────────────────────────────────────────────

export class AgentTracer {
  private readonly maxSpans: number;
  private readonly recordIO: boolean;
  private readonly onSpanComplete: ((span: TraceSpan) => void) | undefined;

  /** Active (running) spans keyed by span ID */
  private activeSpans: Map<string, TraceSpan> = new Map();

  /** Completed spans in insertion order */
  private completedSpans: TraceSpan[] = [];

  constructor(options: AgentTracerOptions = {}) {
    this.maxSpans = options.maxSpans ?? 1000;
    this.recordIO = options.recordIO ?? true;
    this.onSpanComplete = options.onSpanComplete;
  }

  // ── Span Lifecycle ──────────────────────────────────────────────────────────

  /**
   * Create and start a new span.
   *
   * If parentContext is provided the new span is a child of that context's
   * current span and shares the same traceId. Otherwise a new root span
   * (and therefore a new trace) is created.
   *
   * Returns both the span and a TraceContext that callers should propagate
   * to child operations so they can link back via parentId.
   */
  startSpan(
    agentName: string,
    operation: string,
    parentContext?: TraceContext,
    inputs?: Record<string, unknown>,
  ): { span: TraceSpan; context: TraceContext } {
    const id = this.generateId();
    const traceId = parentContext?.traceId ?? this.generateId();
    const parentId = parentContext?.spanId ?? null;

    const span: TraceSpan = {
      id,
      parentId,
      traceId,
      agentName,
      operation,
      startTime: new Date().toISOString(),
      endTime: null,
      durationMs: null,
      status: 'running',
    };

    if (this.recordIO && inputs !== undefined) {
      span.inputs = this.sanitizeIO(inputs);
    }

    if (parentContext?.baggage) {
      span.tags = { ...parentContext.baggage };
    }

    this.activeSpans.set(id, span);

    const context: TraceContext = {
      traceId,
      spanId: id,
      baggage: parentContext?.baggage,
    };

    return { span, context };
  }

  /**
   * Complete a span. Computes duration and transitions status to success or error.
   *
   * Returns the completed span or null if the span ID was not found.
   */
  endSpan(
    spanId: string,
    result?: {
      status: 'success' | 'error';
      outputs?: Record<string, unknown>;
      error?: string;
      dataSlush?: TraceSpan['dataSlush'];
    },
  ): TraceSpan | null {
    const span = this.activeSpans.get(spanId);
    if (!span) return null;

    const endTime = new Date();
    const startMs = new Date(span.startTime).getTime();

    span.endTime = endTime.toISOString();
    span.durationMs = endTime.getTime() - startMs;
    span.status = result?.status ?? 'success';

    if (result?.error !== undefined) {
      span.error = result.error;
    }

    if (this.recordIO && result?.outputs !== undefined) {
      span.outputs = this.sanitizeIO(result.outputs);
    }

    if (result?.dataSlush !== undefined) {
      span.dataSlush = result.dataSlush;
    }

    this.activeSpans.delete(spanId);
    this.addCompleted(span);

    if (this.onSpanComplete) {
      this.onSpanComplete(span);
    }

    return span;
  }

  // ── Querying ────────────────────────────────────────────────────────────────

  /**
   * Retrieve a span by ID, searching both active and completed spans.
   */
  getSpan(spanId: string): TraceSpan | null {
    return this.activeSpans.get(spanId) ?? this.completedSpans.find(s => s.id === spanId) ?? null;
  }

  /**
   * Get all spans (active and completed) belonging to a trace, in chronological order.
   */
  getTrace(traceId: string): TraceSpan[] {
    const active = Array.from(this.activeSpans.values()).filter(s => s.traceId === traceId);
    const completed = this.completedSpans.filter(s => s.traceId === traceId);

    return [...active, ...completed].sort((a, b) => a.startTime.localeCompare(b.startTime));
  }

  /**
   * Get all spans currently in the 'running' state.
   */
  getActiveSpans(): TraceSpan[] {
    return Array.from(this.activeSpans.values());
  }

  /**
   * Get recently completed spans, newest first.
   *
   * @param limit - Maximum number of spans to return. Defaults to all retained spans.
   */
  getCompletedSpans(limit?: number): TraceSpan[] {
    const reversed = [...this.completedSpans].reverse();
    return limit !== undefined ? reversed.slice(0, limit) : reversed;
  }

  // ── Utility ─────────────────────────────────────────────────────────────────

  /**
   * Clear all spans (active and completed). Useful between test runs.
   */
  clear(): void {
    this.activeSpans.clear();
    this.completedSpans = [];
  }

  /**
   * Return a serializable summary of current tracer state, suitable for
   * dashboards or JSON export.
   */
  toJSON(): object {
    const completedByTrace = new Map<string, TraceSpan[]>();
    for (const span of this.completedSpans) {
      const group = completedByTrace.get(span.traceId) ?? [];
      group.push(span);
      completedByTrace.set(span.traceId, group);
    }

    const traces = Array.from(completedByTrace.entries()).map(([traceId, spans]) => {
      const sorted = [...spans].sort((a, b) => a.startTime.localeCompare(b.startTime));
      const root = sorted.find(s => s.parentId === null) ?? sorted[0];
      const errors = spans.filter(s => s.status === 'error');
      const durations = spans.map(s => s.durationMs ?? 0);
      const totalMs = durations.reduce((acc, d) => acc + d, 0);

      return {
        traceId,
        spanCount: spans.length,
        status: errors.length > 0 ? 'error' : 'success',
        rootAgent: root?.agentName ?? null,
        startTime: root?.startTime ?? null,
        totalDurationMs: totalMs,
        errorCount: errors.length,
      };
    });

    return {
      activeSpans: Array.from(this.activeSpans.values()),
      completedSpans: [...this.completedSpans].reverse(),
      traceCount: completedByTrace.size,
      maxSpans: this.maxSpans,
      recordIO: this.recordIO,
      traces: traces.sort((a, b) => (b.startTime ?? '').localeCompare(a.startTime ?? '')),
    };
  }

  // ── Private Helpers ─────────────────────────────────────────────────────────

  /**
   * Generate a 16-character ID by truncating a UUID (removing dashes).
   */
  private generateId(): string {
    return randomUUID().replace(/-/g, '').slice(0, 16);
  }

  /**
   * Sanitize an IO record for safe storage: remove private/context keys and
   * truncate large string values so spans stay lightweight.
   */
  private sanitizeIO(io: Record<string, unknown>): Record<string, unknown> {
    const sanitized: Record<string, unknown> = {};

    for (const [key, value] of Object.entries(io)) {
      // Skip internal agent context injected by execute()
      if (key === '_context' || key === '_sloshFilter' || key === '_sloshPreferences') {
        continue;
      }

      if (typeof value === 'string' && value.length > 500) {
        sanitized[key] = value.slice(0, 500) + '…';
      } else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
        // Shallow clone nested objects without recursing into large trees
        sanitized[key] = '[object]';
      } else {
        sanitized[key] = value;
      }
    }

    return sanitized;
  }

  /**
   * Append a completed span and enforce the maxSpans retention limit.
   * Oldest spans are evicted first (FIFO).
   */
  private addCompleted(span: TraceSpan): void {
    this.completedSpans.push(span);
    if (this.completedSpans.length > this.maxSpans) {
      this.completedSpans.splice(0, this.completedSpans.length - this.maxSpans);
    }
  }
}

// ── Factory & Singleton ───────────────────────────────────────────────────────

/**
 * Create a new AgentTracer with the given options.
 */
export function createTracer(options?: AgentTracerOptions): AgentTracer {
  return new AgentTracer(options);
}

/**
 * Global tracer singleton. Ready to use with zero configuration.
 * Override by replacing with a custom instance from `createTracer()`.
 */
export const globalTracer: AgentTracer = createTracer();
